api/pkg/inference-logger/logger/mlobs_sink.go

```diff
@@ -0,0 +1,154 @@
+}
+
+func NewMLObsSink(
+	logger *zap.SugaredLogger,
+	producer KafkaProducer,
+	projectName string,
+	modelName string,
+	modelVersion string,
+) LogSink {
+	sink := &MLObsSink{
+		logger:       logger,
+		producer:     producer,
+		projectName:  projectName,
+		modelName:    modelName,
+		modelVersion: modelVersion,
+	}
+	go sink.handleMessageDelivery()
+
+	return sink
+}
+
+func (m *MLObsSink) newPredictionLog(rawLogEntry *LogEntry) (*upiv1.PredictionLog, error) {
+	predictionLog := &upiv1.PredictionLog{}
+	standardModelRequest := &StandardModelRequest{}
+	err := json.Unmarshal(rawLogEntry.RequestPayload.Body, standardModelRequest)
+	if err != nil {
+		return nil, err
+	}
+	standardModelResponse := &StandardModelResponse{}
+	err = json.Unmarshal(rawLogEntry.ResponsePayload.Body, standardModelResponse)
+	if err != nil {
+		return nil, err
+	}
+	predictionLog.PredictionId = standardModelRequest.SessionId
+	predictionLog.ModelName = m.modelName
+	predictionLog.ModelVersion = m.modelVersion
+	predictionLog.ProjectName = m.projectName
+	featuresTable := &structpb.Struct{
+		Fields: make(map[string]*structpb.Value),
+	}
+	predictionTable := &structpb.Struct{
+		Fields: make(map[string]*structpb.Value),
+	}
+	featureTableData := &structpb.ListValue{
+		Values: make([]*structpb.Value, len(standardModelRequest.Instances)),
+	}
+	predictionTableData := &structpb.ListValue{
+		Values: make([]*structpb.Value, len(standardModelResponse.Predictions)),
+	}
+	rowIds := &structpb.ListValue{
+		Values: make([]*structpb.Value, len(standardModelRequest.RowIds)),
+	}
+	for i, instance := range standardModelRequest.Instances {
+		featureTableRow := &structpb.ListValue{
+			Values: make([]*structpb.Value, len(instance)),
+		}
+		for j, value := range instance {
+			if value != nil {
+				featureTableRow.Values[j] = structpb.NewNumberValue(*value)
+			} else {
+				featureTableRow.Values[j] = structpb.NewNullValue()
+			}
+		}
+
+		featureTableData.Values[i] = structpb.NewListValue(featureTableRow)
+		predictionTableData.Values[i] = structpb.NewNumberValue(standardModelResponse.Predictions[i])
+		rowIds.Values[i] = structpb.NewStringValue(standardModelRequest.RowIds[i])
+	}
+	featuresTable.Fields["data"] = structpb.NewListValue(featureTableData)
+	featuresTable.Fields["row_ids"] = structpb.NewListValue(rowIds)
+	predictionTable.Fields["data"] = structpb.NewListValue(predictionTableData)
+	predictionTable.Fields["row_ids"] = structpb.NewListValue(rowIds)
+
+	predictionLog.Input = &upiv1.ModelInput{
+		FeaturesTable: featuresTable,
+	}
+	predictionLog.Output = &upiv1.ModelOutput{
+		PredictionResultsTable: predictionTable,
+	}
+	return predictionLog, nil
+}
+
+func (m *MLObsSink) buildNewKafkaMessage(predictionLog *upiv1.PredictionLog) (*kafka.Message, error) {
+	logBytes, err := proto.Marshal(predictionLog)
+	if err != nil {
+		return nil, err
+	}
+	topic := fmt.Sprintf("caraml-%s-%s-%s-prediction-log", m.projectName, m.modelName, m.modelVersion)
```

Comment:
@khorshuheng, can you remind us again why we need one topic per model version instead of one topic per model?

Reply:
Currently, we are deploying one Kafka consumer per model version. Since a deployment is unique to a model version, Merlin API is able to provide the model schema (eg. feature types, prediction schema) to the deployment as a configuration.

If we use one topic per model, instead of version, then the consumer / worker needs to handle the scenario where multiple versions of prediction log are present in the same Kafka topic. In this scenario, the deployment will need to know the model schema for all the model versions, which is not going to be easy to maintain in the form of deployment configuration. Realistically, the deployment will need to call Merlin API instead to fetch and cache the schema for a specific model version.

Reply:
Got it. Thanks for clear explanation!